pyfstat/snr.py

```diff
@@ -0,0 +1,423 @@
+from attrs import define, field
+import lal
+import lalpulsar
+import logging
+import numpy as np
+from typing import Union
+
+from pyfstat.helper_functions import get_ephemeris_files
+
+
+@define(kw_only=True, slots=False)
+class SignalToNoiseRatio:
+    """
+    Compute the optimal signal-to-noise ratio (and derived quantities, such as the F-statistic)
+    of a CW signal on Gaussian noise or an arbitrary dataset of SFTs.
+
+    The definition of SNR (shortcut for "optimal signal-to-noise ratio")
+    is taken from Eq. (76) of https://dcc.ligo.org/T0900149-v6/public and is
+    such that `<2F> = 4 + SNR^2`, where `<2F>` represents the expected
+    value over noise realizations of twice the F-statistic of a perfectly
+    matched template to an existing signal in the data.
+
+    Computing SNR^2 requires two quantities:
+        - The antenna pattern matrix `M`, which depends on the sky position and polarization angle
+        and encodes the effect of the detector's antenna pattern over the course
+        of the observing run.
+        - The JKS amplitude parameters `{A^0, A^1, A^2, A^3}`, which are functions
+        of the CW's amplitude parameters `(h0, cosi, psi)` or, alternatively,
+        `(aPlus, aCross, psi)`.
```

Comment:
Looks like these two comments from initial review fell through the cracks:

- [ ]  Also phi0, though it doesn't really make a difference?

- [ ] Also add JKS ref?

Reply:
```suggestion
        - The JKS amplitude parameters `{A^0, A^1, A^2, A^3}` 
        [Jaranowski, Krolak, Schutz Phys. Rev. D 58 063001,(1998)], which are functions
        of the CW's amplitude parameters `(h0, cosi, psi, phi0)` or, alternatively,
        `(aPlus, aCross, psi, phi0)`.
```